### PR TITLE
Store interface configuration to flash only where documented and implement esp.save_config() to save configuration to flash on demand

### PR DESCRIPTION
**Pull request by kfricke**
Connecting to a network and altering interface modes do also not write the configuration to flash anymore.
This previous behavior was not documented in the MicroPython documentation and aligns the implementation to this.

Setting network configuration parameters using the method `WLAN.config(kwarg=...)` does still write the settings to flash. This behavior is described in the documentation now, containing a warning about flash wearing.

These changes do not contain changes to the network API! The behavior of the implementation prior to this pull request is only known by experienced users of the C-API which had this behavior from the beginning and later got enhanced by API calls that did avoid this (several API methods got a `_current` variant, which does not write config to flash anymore).

### CODE REVIEW

**Comment by kfricke** [comment]
Just as an additional note... Reading and re-setting the MAC address in station mode does implement an undocumented way of storing the station configuration in flash. This is the only way to do this after this patch.

**Comment by pfalcon** [comment]
> This is the only way to do this after this patch.

And that's one, but not the only problem with it. The biggest problem is that it change well known behavior of ESP8266 module. Not just behavior of MicroPython's ESP8266 port how it was from the beginning (familiar to a thousand of users), but the well-known ESP8266 behavior, familiar to hundreds of thousands of ESP8266 users. Everyone who uses ESP8266 knows that once they called connect() function, the module remembers connection and will automatically connect after reset, power down/power up, etc. Not only that, that mode of operation is familiar to all users of desktop and mobile computer. Like, on your phone, you don't need to type access point name and password each time you press power button, it's already connected when you do that. That's behavior familiar to billions of people. Ain't it a bit too ambitious to try to change that?

**Comment by kfricke** [comment]
>>This is the only way to do this after this patch.

>And that's one, but not the only problem with it. ...

This got changed with the last three commits (). It is now possible to save the config consistently with `esp.save_config()`.

> The biggest problem is that it change well known behavior of ESP8266 module. Not just behavior of MicroPython's ESP8266 port how it was from the beginning (familiar to a thousand of users), but the well-known ESP8266 behavior, familiar to hundreds of thousands of ESP8266 users. Everyone who uses ESP8266 knows that once they called connect() function, the module remembers connection and will automatically connect after reset, power down/power up, etc.

People come to MicroPython not because they are familiar with the ESP8266. They are here because of Python. 
Let us look at the situation from another point of view: We have not documented the current behavior because it breaks the API if we tell the truth. If we'd align the documentation we would notice immediately that the vocabulary of the API methods is not sufficiently reflecting what the methods actually do.

> ...that mode of operation is familiar to all users of desktop and mobile computer... phone.

There is not a human in the ESP8266 and this is a micro-controller. In our case we have users who want to write Python scripts which tell the micro step by step what to do.

Do not compare an apple with a rice corn!
_(rice was chosen because of it's size not it's origin from asia ;) )_

>Ain't it a bit too ambitious to try to change that?

Not if it the behavior does destroy flash in concrete and realistic use-cases. Also "well-known" and undocumented does not really stand the truth.

Ain't yours a bad attitude to support progress?

**Comment by pfalcon** [comment]
> Ain't yours a bad attitude to support progress?

I'm sorry then, and leave your issue(s) for somebody else to communicate with.

**Comment by pfalcon** [comment]
To get forward with this, I suggest posting an announcement on the forum, explaining fairly and straightforwardly what you try to achieve - and that's breaking how existing way of doing things. Explain why you do this, explain the issues with this approach (one would need to put AP password in plaintext in boot.py!), then explain what you do to address that. Attach voting to the thread to see what people think. Then in couple of releases we can go for this. If the above was done right away, we'd already be some way thru it.

**Comment by kfricke** [comment]
Done.
http://forum.micropython.org/viewtopic.php?f=16&t=2631

**Comment by puuu** [comment]
Sorry for not writing at the forum, but after reading the thread I can not see any step forward nor any productive discussion.

Let's be more objective. @kfricke sometimes a pull request needs time, please be relaxed.

First I want to state. I like the behaviour of the ESP8266, that automatic connects after reboot. It provides a fantastic user experience. But the fact, that `w.connect(...)` and `w.active(...)` rewrites the flash and lead flash wear-out is a no go. I also run some ESP8266 on battery that is why I am interested in a fast solution of this issue.

Some thoughts to arguments that popped up against this pull request:

> Explain why you do this, explain the issues with this approach (one would need to put AP password in plaintext in boot.py!), then explain what you do to address that.

@pfalcon merged my pull request #2506. So there is no need to deal with the password in any script.

Just save the credentials use during setup time:
```python
sta_if = network.WLAN(network.STA_IF)
sta_if.active(True)
sta_if.connect("<AP_name>", "<password>"
esp.save_config()
```

Then you can run on every start:
```python
sta_if = network.WLAN(network.STA_IF)
sta_if.active(True)
sta_if.connect()
```
That is all, no WiFi credentials!

> The biggest problem is that it change well known behavior of ESP8266 module. Not just behavior of MicroPython's ESP8266 port how it was from the beginning (familiar to a thousand of users), but the well-known ESP8266 behavior, familiar to hundreds of thousands of ESP8266 users.

That is true and I don't like it, because it breaks the user experience. But why not saving `opmode` in `esp_save_config()`? I would suggest adding something like `wifi_set_opmode(wifi_get_opmode());` in `esp_save_config()`.

This way you do not have to change anything in the user scripts, because the ESP8266 automatic connect during boot time. Only during setup time you have to run `esp.save_config()` in addition (I believe that every MicroPython user can deal with that, because it is less annoying than the change of ticks_diff()).

If you don't want any automatic connect during boot time (e.g. battery driven), just run once: `sta_if.active(False); esp.save_config()`.

I am not sure if `esp.save_config()` is the right name, maybe `esp.save_wifi_config()` is better, or something else maybe under machine as official API...

**Comment by kfricke** [comment]
> I am not sure if esp.save_config() is the right name, maybe esp.save_wifi_config() is better, or something else maybe under machine as official API...

I have thought of this as well, but all the ESP SDK does not make that destinction, so i did.

> ... it breaks the user experience. ...

Okay of course it does change things. But does it really kill the user experience!? All this PR does is let the `wlan.connect(...)` and `active(...)` methods act in non persistent manner and avoid flash writes. In case you want to persist settings call `esp.save_config()` and after the next restart the ESP and it will reconnect as you like it to do so!

**Comment by kfricke** [comment]
> That is all, no WiFi credentials!

As mentioned a lot of times by many people: Storing the Pre-Shared-Key to a wireless network in plain-text is common and best practice. We must not think it is secret. It simply is not public and that is the way we store it.

It is the same way as Espressif does store it, just at another place. And storing it in front of the users eyes it better than somewhere hidden.

 _"There is no security by obscurity!"_

File: esp8266/modesp.c
Comment (puuu) [inline esp8266/modesp.c]: perhaps, better would be to use `error_check()`
`error_check(wifi_station_get_config(&config), "can't get STA config");`

File: esp8266/modesp.c
Comment (puuu) [inline esp8266/modesp.c]: `error_check(wifi_station_set_config(&config), "can't save STA config");`

File: esp8266/modesp.c
Comment (puuu) [inline esp8266/modesp.c]: `error_check(wifi_softap_get_config(&config), "can't get AP config");`

File: esp8266/modesp.c
Comment (puuu) [inline esp8266/modesp.c]: `error_check(wifi_softap_set_config(&config), "can't save AP config");`

File: esp8266/modesp.c
Comment (puuu) [inline esp8266/modesp.c]: don't forget to save the interface status!
`error_check(wifi_set_opmode(wifi_get_opmode()), "can't save i/f status");`

**Review by puuu — CHANGES_REQUESTED** [review]

**Comment by puuu** [comment]
> In case you want to persist settings call esp.save_config() and after the next restart the ESP and it will reconnect as you like it to do so!

Not yet! You also have to save the interface status . Then I think, this is a acceptable behavior.

With respect to other suggestions, I like your idea of having a function that only saves the WiFi configuration, because of simplicity. E.g., you don't have to deal with internal states.

I tested it with my suggested changes and it runs fine.

Please update the pull request and rebase it, because of the conflict in `esp8266/modesp.c`.

>> That is all, no WiFi credentials!

> As mentioned a lot of times by many people: Storing the Pre-Shared-Key to a wireless network in plain-text is common and best practice. We must not think it is secret. It simply is not public and that is the way we store it.

#2506 was not mainly because of security.  Not to deal with WiFi credentials in your scripts enhance the user experience and reduce the memory usage (WiFi PSK can be long).

**Comment by puuu** [comment]
>> I am not sure if esp.save_config() is the right name, maybe esp.save_wifi_config() is better, or something else maybe under machine as official API...

> I have thought of this as well, but all the ESP SDK does not make that destinction, so i did.

Indeed the name is the most critical part of this pull request, because it hurts any user. We can not suggest any name and change it later again.

So we have to discuss is in a constructive way. Here we also require the experience of the maintainers. As @palcon already said, on many devices the password is saved. Thus, we have to expect that the ESP8266 will not be the only device that has the possibility to save the WiFi credentials. This means we should try to describe an official API function, that also can be used by further boards.

**Comment by kfricke** [comment]
I will try to find time tomorrow evening. But of course this PR will hang around for some time before being merged, so no hurry. 

Are the `error_check(...)` wrappings really needed? The config structs are filled by API calls and not user supplied inputs, so my assumption was to be allowed to skip it there.

**Comment by kfricke** [comment]
I did rename the method `esp.save_config()` to `esp.wifi_save_config()` after introducing the new method `esp.wifi_station_auto_connect(...)`. This is to follow loosely the naming scheme of the methods found in the ESP SDK, so everyone can feel familiar with it.
The later lets you set and get the behavior of automatic connect on startup and got mentioned by me in #2620

**Comment by kfricke** [comment]
By the way, it seems that (by reading the ESP SDK docs), that only the wireless network interface configuration parameters are being written to the 16K segment of flash. Maybe the client certificates for the corresponding WPA2 authentication (aka "WPA2 Enterprise") are put there as well, but this still is network configuration.

**Comment by kfricke** [comment]
Bye

**Pull request closed without merging**
